commit 9f0ffa418483938d25a15f6ad3891389f333bc59
Author: Rohan Dutta <drohan@codeaurora.org>
Date:   Tue Oct 27 12:09:10 2020 +0200

    cfg80211: Add support to configure SAE PWE value to drivers
    
    Add support to configure SAE PWE preference from userspace to drivers in
    both AP and STA modes. This is needed for cases where the driver takes
    care of Authentication frame processing (SME in the driver) so that
    correct enforcement of the acceptable PWE derivation mechanism can be
    performed.
    
    The userspace applications can pass the sae_pwe value using the
    NL80211_ATTR_SAE_PWE attribute in the NL80211_CMD_CONNECT and
    NL80211_CMD_START_AP commands to the driver. This allows selection
    between the hunting-and-pecking loop and hash-to-element options for PWE
    derivation. For backwards compatibility, this new attribute is optional
    and if not included, the driver is notified of the value being
    unspecified.
    
    Signed-off-by: Rohan Dutta <drohan@codeaurora.org>
    Signed-off-by: Jouni Malinen <jouni@codeaurora.org>
    Link: https://lore.kernel.org/r/20201027100910.22283-1-jouni@codeaurora.org
    Signed-off-by: Johannes Berg <johannes.berg@intel.com>

diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 554796a6c6fe..0928ecbe5bd6 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -715,6 +715,9 @@ static const struct nla_policy nl80211_policy[NUM_NL80211_ATTR] = {
 		NLA_POLICY_EXACT_LEN(IEEE80211_S1G_CAPABILITY_LEN),
 	[NL80211_ATTR_S1G_CAPABILITY_MASK] =
 		NLA_POLICY_EXACT_LEN(IEEE80211_S1G_CAPABILITY_LEN),
+	[NL80211_ATTR_SAE_PWE] =
+		NLA_POLICY_RANGE(NLA_U8, NL80211_SAE_PWE_HUNT_AND_PECK,
+				 NL80211_SAE_PWE_BOTH),
 };
 
 /* policy for the key attributes */
@@ -9731,6 +9734,12 @@ static int nl80211_crypto_settings(struct cfg80211_registered_device *rdev,
 			nla_len(info->attrs[NL80211_ATTR_SAE_PASSWORD]);
 	}
 
+	if (info->attrs[NL80211_ATTR_SAE_PWE])
+		settings->sae_pwe =
+			nla_get_u8(info->attrs[NL80211_ATTR_SAE_PWE]);
+	else
+		settings->sae_pwe = NL80211_SAE_PWE_UNSPECIFIED;
+
 	return 0;
 }